File: SampleApps /ObjC/SampleAppWakeUp/SampleAppWakeUp/Services/NotificationService/Notification.m
//
//  Notification.m
//  SampleAppWakeUp
//
//  Created by Akhil Rana on 26/11/23.
//

#import "Notification.h"
@import UIKit;
@import UserNotifications;

@implementation Notification

+ (void)scheduleLocalNotificationWithTitle:(NSString *)title subTitle:(NSString *)subTitle body:(NSString *)body {
    // Request permission to show notifications
    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
    [center requestAuthorizationWithOptions:(UNAuthorizationOptionAlert + UNAuthorizationOptionSound)
                          completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if (granted) {
            // Permission granted, create and schedule the notification
            [self createAndScheduleNotificationWithTitle:title subTitle:subTitle body:body];
        } else {
            // Handle the case where the user denied permission
            NSLog(@"Permission denied for notifications");
        }
    }];
}

+ (void)createAndScheduleNotificationWithTitle:(NSString *)title subTitle:(NSString *)subTitle body:(NSString *)body {
    // Create notification content
    UNMutableNotificationContent *content = [[UNMutableNotificationContent alloc] init];
    content.title = title;
    content.subtitle = subTitle;
    content.body = body;
    content.sound = [UNNotificationSound defaultSound];

    // Set up trigger for 10 seconds from now
    NSTimeInterval timeInterval = 1;
    UNTimeIntervalNotificationTrigger *trigger = [UNTimeIntervalNotificationTrigger triggerWithTimeInterval:timeInterval repeats:NO];

    // Create notification request
    NSString *identifier = [[NSUUID UUID] UUIDString];
    UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:identifier
                                                                          content:content
                                                                          trigger:trigger];

    // Add the request to the notification center
    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
    [center addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
        if (error != nil) {
            NSLog(@"Error scheduling notification: %@", error.localizedDescription);
        } else {
            NSLog(@"Notification scheduled successfully!");
        }
    }];
}

@end

File: SampleApps /SampleBlueDotWithIndoorLocation/SampleBlueDotWithIndoorLocation/Utilities/Notification.swift
//
//  Notification.swift
//  SampleBlueDotWithIndoorLocation
//
//  Created by Rajesh Vishwakarma on 16/03/23.
//

import UIKit
import UserNotifications

struct NotificationData {
    let title: String
    let subtitle: String
    let body: String
}

extension Notification {
    
    static func requestNotificationPermission() {
        let options: UNAuthorizationOptions = [.badge, .sound, .alert]
        UNUserNotificationCenter.current()
            .requestAuthorization(options: options) { _, error in
                if let error = error {
                    debugPrint("Error: \(error)")
                }
            }
    }
    
    static func schedule(after time: TimeInterval, payload: NotificationData) {
        let content = UNMutableNotificationContent()
        content.title = payload.title
        content.subtitle = payload.subtitle
        content.body = payload.body
        content.badge = 1
        
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: time, repeats: false)
        let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: trigger)
        
        UNUserNotificationCenter.current().add(request, withCompletionHandler: nil)
    }
    
    static func resetNotificationBadeCount() {
        UIApplication.shared.applicationIconBadgeNumber = 0
        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
        UNUserNotificationCenter.current().removeAllDeliveredNotifications()
    }
}
